ash/src/lib.rs

```diff
@@ -66,6 +66,40 @@ impl<'r, T> RawPtr<T> for Option<&'r T> {
     }
 }
 
+/// Given a mutable raw pointer to a type with an `s_type` member such as [`vk::BaseOutStructure`],
+/// match on a set of Vulkan structures. The struct will be rebound to the
+/// given variable of the type of the given Vulkan structure.
+///
+/// ```
+/// let mut info = ash::vk::DeviceCreateInfo::default();
+/// let info: *mut ash::vk::BaseOutStructure = <*mut _>::cast(&mut info);
+/// unsafe {
+///     ash::match_struct!(match info {
+///         info @ ash::vk::DeviceQueueCreateInfo => {
+///             dbg!(&info); // Unreachable
+///         }
+///         info @ ash::vk::DeviceCreateInfo => {
+///             dbg!(&info);
+///         }
+///     })
+/// }
+/// ```
+#[macro_export]
+macro_rules! match_struct {
+    (match $p:ident { $($bind:ident @ $ty:path => $body:block $(,)?)+ }) => {
+        match (*$p).s_type {
+            $(<$ty as $crate::vk::HasStructureType>::STRUCTURE_TYPE => {
+                let $bind = $p
+                    .cast::<$ty>()
+                    .as_mut()
```

Comment:
Only mutable sturctures work with this, currently :/

Reply:
Maybe separate `match_out_struct` and `match_in_struct` macros?

Reply:
I like those names!

Just unfortunate that it's going to be ~23 lines of duplication for the entire macro and doc-comment.

Reply:
Yeah; as far as compromises to deal with declarative macros being kind of crappy go, it could be a lot worse.